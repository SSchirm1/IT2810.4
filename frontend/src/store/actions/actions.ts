import {
  GET_CITIES,
  GET_CITY,
  GET_STUDENTCITIES,
  GET_STUDENTCITY,
  apiActionTypes
} from "./actiontypes";
import axios from "axios";
import { ThunkAction } from "redux-thunk";
import { Dispatch } from "redux";
import { City, StudentCity } from "../interfaces";

import Filter from "../../hooks/Filter/interfaces";

const API = process.env.REACT_APP_API_SERVER ?? "http://it2810-72.idi.ntnu.no/api";
const OFFSET = 4;

export function GetCities(): ThunkAction<
  Promise<void>,
  {},
  {},
  apiActionTypes
> {
  console.log("GetCities");
  return async (dispatch: Dispatch<apiActionTypes>) => {
    axios.get(API + "/byer").then(res => {
      const cities: City[] = res.data;
      dispatch({
        type: GET_CITIES,
        cities: cities
      });
    });
  };
}

export function GetStudentCities(
  page: number,
  filter: Filter
): ThunkAction<Promise<void>, {}, {}, apiActionTypes> {
  console.log("GetStudentCities");

  return async (dispatch: Dispatch<apiActionTypes>) => {
    axios
      .get(`${API}/studentbyer`, {
        params: {
          take: OFFSET,
          skip: page * OFFSET,
          sort: filter.sort,
          querystring: filter.queryString,
          filter: filter.city
        }
      })
      .then(res => {
        console.log(res.data);
        const studentCities: StudentCity[] = res.data.studentbyer;
        const count: number = res.data.count;
        dispatch({
          type: GET_STUDENTCITIES,
          studentCities: studentCities,
          count: count
        });
      });
  };
}

export function GetStudentCity(
  id: number
): ThunkAction<Promise<void>, {}, {}, apiActionTypes> {
  console.log("GetStudentCity");
  return async (dispatch: Dispatch<apiActionTypes>) => {
    axios.get(API + "/studentbyer/" + id.toString()).then(res => {
      //TODO: change '1' to input value
      const studentcity: StudentCity = res.data;
      dispatch({
        type: GET_STUDENTCITY,
        studentCity: studentcity
      });
    });
  };
}

export function GetCity(
  id: number
): ThunkAction<Promise<void>, {}, {}, apiActionTypes> {
  console.log("GetCity");
  return async (dispatch: Dispatch<apiActionTypes>) => {
    axios.get(API + "/byer/" + id.toString()).then(res => {
      //TODO: change '1' to input value
      const city: City = res.data;
      dispatch({
        type: GET_CITY,
        city: city
      });
    });
  };
}
