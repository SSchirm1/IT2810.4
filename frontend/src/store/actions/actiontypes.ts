import { StudentCity, City } from "../interfaces";
import Filter from "./interfaces";

export const GET_STUDENTCITIES = "GET_STUDENTCITIES";
export const GET_STUDENTCITY = "GET_STUDENTCITY";

export const GET_CITIES = "GET_CITIES";
export const GET_CITY = "GET_CITY";
export const SET_FILTER = "SET_FILTER";

interface GetStudentCityAction {
  type: typeof GET_STUDENTCITY;
  studentCity: StudentCity;
}

interface GetStudentCitiesAction {
  type: typeof GET_STUDENTCITIES;
  studentCities: StudentCity[];
  count: number;
}

export interface GetCitiesAction {
  type: typeof GET_CITIES;
  cities: City[];
}
interface GetCityAction {
  type: typeof GET_CITY;
  city: City;
}

export interface SetFilterAction {
  type: typeof SET_FILTER;
  filter: Filter;
}

export type apiActionTypes =
  | GetStudentCitiesAction
  | GetStudentCityAction
  | GetCitiesAction
  | GetCityAction;
