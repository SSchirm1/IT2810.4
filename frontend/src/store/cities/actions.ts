import { City } from "./interfaces";
import {
  CitiesAction,
  FAILURE_CITIES,
  FETCH_CITIES,
  PENDING_CITIES,
  SUCCESS_CITIES,
} from "./actionTypes";

export function fetchCities(): CitiesAction {
  return {
    type: FETCH_CITIES,
  };
}
export function pendingCities(): CitiesAction {
  return {
    type: PENDING_CITIES,
  };
}
export function successCities(data: City[]): CitiesAction {
  return {
    type: SUCCESS_CITIES,
    cities: data,
  };
}
export function failureCities(): CitiesAction {
  return {
    type: FAILURE_CITIES,
  };
}
