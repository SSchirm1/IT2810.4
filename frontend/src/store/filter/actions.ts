import { SetFilterAction, SET_FILTER } from "./actionTypes";

import Filter from "./interfaces";

export function setFilter(filter: Filter): SetFilterAction {
  return {
    type: SET_FILTER,
    filter
  };
}
