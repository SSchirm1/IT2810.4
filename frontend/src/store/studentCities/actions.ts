import { StudentCity } from "./interfaces";
import {
  FAILURE_STUDENT_CITIES,
  FETCH_STUDENT_CITIES,
  PENDING_STUDENT_CITIES,
  SUCCESS_STUDENT_CITIES,
  StudentCitiesAction,
} from "./actionTypes";

export function fetchStudentCities(): StudentCitiesAction {
  return {
    type: FETCH_STUDENT_CITIES,
  };
}
export function pendingStudentCities(): StudentCitiesAction {
  return {
    type: PENDING_STUDENT_CITIES,
  };
}
export function successStudentCities(
  data: StudentCity[],
  count: number
): StudentCitiesAction {
  return {
    type: SUCCESS_STUDENT_CITIES,
    studentCities: data,
    count,
  };
}
export function failureStudentCities(): StudentCitiesAction {
  return {
    type: FAILURE_STUDENT_CITIES,
  };
}
