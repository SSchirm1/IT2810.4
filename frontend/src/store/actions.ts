import {SET_SORTING, SET_FILTER, SET_SEARCH, ActionTypes, filter, sort, search} from './actiontypes'

export function setSorting(sortby: sort): ActionTypes {
    return {
      type: SET_SORTING,
      sort: sortby
    }
  }

export function setFilter(filterby: filter): ActionTypes {
    return {
        type: SET_FILTER,
        filter: filterby
    }
  }

export function setSearch(searchafter: search): ActionTypes {
    return {
        type: SET_SEARCH,
        search: searchafter
    }
  }