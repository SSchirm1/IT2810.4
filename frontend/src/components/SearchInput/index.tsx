import React, { useEffect } from "react";
import { Box, useColorMode, Select, Input } from "@chakra-ui/core";
import { useActions } from "../../hooks/useActions";
import { RootState } from "../../store/reducers";
import { useSelector } from "react-redux";
import { Sort } from "../../store/actions/interfaces";
import { setFilter, fetchCities } from "../../store/actions/actions";

function Search_parameters() {
  const { colorMode } = useColorMode();
  const { cities } = useSelector((state: RootState) => {
    return {
      cities: state.cities.cities
    };
  });

  const actions = useActions({ setFilter, fetchCities });
  const filter = useSelector((state: RootState) => state.filter.filter);
  const [value, setValue] = React.useState("");

  useEffect(() => {
    //TODO: kanskje denne burde blitt gjort et annet sted?
    updateSort("alphabetical");
    actions.fetchCities();
  }, []);

  const updateSort = (value: Sort) => {
    actions.setFilter({ ...filter, sort: value, page: 0 });
    //actions.GetStudentCities({ ...filter, sort: value, page: 0 });
  };
  const updateCity = (value: string) => {
    actions.setFilter({ ...filter, city: value, page: 0 });
    //actions.GetStudentCities({ ...filter, city: value, page: 0 });
  };

  const handleSearch = (value: string) => {
    actions.setFilter({
      ...filter,
      queryString: value,
      page: 0
    });
    //actions.GetStudentCities({ ...filter, queryString: value, page: 0 });

    setValue(value);
  };
  const currentCities = cities.phase === "SUCCESS" ? cities.data ?? [] : [];

  return (
    <Box
      width="full"
      marginTop="10px"
      marginBottom="10px"
      p={3}
      boxShadow="md"
      rounded="lg"
      bg={colorMode === "light" ? "gray.100" : "gray.700"}
      color={colorMode === "light" ? "gray.900" : "gray.100"}
    >
      <Input
        _placeholder={{ color: "gray.500" }}
        variant="outline"
        value={value}
        onChange={(event: any) => handleSearch(event.target.value)}
        placeholder="søk etter studentby"
        marginBottom="5px"
        bg={colorMode === "light" ? "white" : "gray.700"}
      />
      <Select
        onChange={event => updateCity(event.currentTarget.value)}
        marginBottom="5px"
        bg={colorMode === "light" ? "white" : "gray.700"}
      >
        <option value="">Alle byer</option>
        {currentCities.map(city => {
          return (
            <option key={city.id} value={city.id}>
              {city.navn}
            </option>
          );
        })}
      </Select>
      <Select
        onChange={event => updateSort(event.currentTarget.value as Sort)}
        bg={colorMode === "light" ? "white" : "gray.700"}
      >
        <option value="alphabetical">{"Alfabetisk A -> Å"}</option>
        <option value="inverseAlphabetical">{"Alfabetisk Å -> A"}</option>
        <option value="ratingHighToLow">{"Total vurdering høy -> lav"}</option>
        <option value="ratingLowToHigh">{"Total vurdering lav -> høy"}</option>
      </Select>
    </Box>
  );
}
export default Search_parameters;
