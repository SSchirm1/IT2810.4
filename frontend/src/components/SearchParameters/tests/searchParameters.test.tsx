/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, fireEvent, screen } from "./testUtils";
import "@testing-library/jest-dom/extend-expect";
import SearchInput from "..";

import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { Store, AnyAction } from "redux";
import { RenderResult } from "@testing-library/react";
import { fetchCities, setFilter } from "../../../store/actions/actions";

const mockStore = configureStore([]);
describe("My Connected React-Redux Component", () => {
  let store: Store<any, AnyAction>;
  let component: RenderResult<typeof import("@testing-library/dom/queries")>;

  beforeEach(() => {
    store = mockStore({
      cities: {
        cities: {
          phase: "SUCCESS",
          data: [
            {
              id: 1,
              navn: "Oslo",
            },
            {
              id: 2,
              navn: "Trondheim",
            },
          ],
          count: null,
        },
      },
      filter: {
        filter: {
          sort: "alphabetical",
          queryString: "",
          page: 0,
          city: "",
        },
      },
    });
    store.dispatch = jest.fn();
    component = render(
      <Provider store={store}>
        <SearchInput />
      </Provider>,
      {}
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  it("renders with correct cities as options", async () => {
    expect(screen.getByRole("option", { name: "Oslo" })).toHaveAttribute(
      "value",
      "1"
    );

    expect(screen.getByRole("option", { name: "Trondheim" })).toHaveAttribute(
      "value",
      "2"
    );
  });

  it("input renders with correct placeholder", async () => {
    expect(screen.getByRole("textbox")).toHaveAttribute(
      "placeholder",
      "sÃ¸k etter studentby"
    );
  });

  it("dispatches action when searching", async () => {
    fireEvent.change(screen.getByRole("textbox"), { target: { value: "a" } });
    expect(store.dispatch).toHaveBeenCalledWith(
      setFilter({ sort: "alphabetical", queryString: "a", page: 0, city: "" })
    );
  });

  it("dispatches action when selecting city", async () => {
    fireEvent.change(screen.getByTestId("select-cities"), {
      target: { value: "1" },
    });
    expect(store.dispatch).toHaveBeenCalledWith(
      setFilter({
        sort: "alphabetical",
        queryString: "",
        page: 0,
        city: "1",
      })
    );
  });

  it("dispatches action when selecting sort", async () => {
    fireEvent.change(screen.getByTestId("select-sort"), {
      target: { value: "inverseAlphabetical" },
    });
    expect(store.dispatch).toHaveBeenCalledWith(
      setFilter({
        sort: "inverseAlphabetical",
        queryString: "",
        page: 0,
        city: "",
      })
    );
  });
});
