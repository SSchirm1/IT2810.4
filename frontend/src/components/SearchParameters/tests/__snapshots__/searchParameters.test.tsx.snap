// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<DocumentFragment>
  <div
    class="css-aw16qv"
  >
    <input
      class="chakra-input css-ttucp5"
      placeholder="søk etter studentby"
      value=""
    />
    <div
      class="chakra-select__wrapper css-9vauj5"
    >
      <select
        class="chakra-select css-1msa8b1"
        data-testid="select-cities"
      >
        <option
          value=""
        >
          Alle byer
        </option>
        <option
          value="1"
        >
          Oslo
        </option>
        <option
          value="2"
        >
          Trondheim
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-pezv7i"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="chakra-select__wrapper css-42b2qy"
    >
      <select
        class="chakra-select css-1msa8b1"
        data-testid="select-sort"
      >
        <option
          value="alphabetical"
        >
          Alfabetisk A -&gt; Å
        </option>
        <option
          value="inverseAlphabetical"
        >
          Alfabetisk Å -&gt; A
        </option>
        <option
          value="ratingHighToLow"
        >
          Total vurdering høy -&gt; lav
        </option>
        <option
          value="ratingLowToHigh"
        >
          Total vurdering lav -&gt; høy
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-pezv7i"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
